gpii/node_modules/solutionsRegistry/src/js/filtered-settings-middleware.js

```diff
@@ -0,0 +1,174 @@
+
+    // Filter the underlying schema using the derived settings.
+    var filteredSchema = gpii.universal.solutionsRegistry.filterSchemaToSettings(observedSettings, baseSchema);
+    return filteredSchema;
+};
+
+fluid.defaults("gpii.universal.solutionsRegistry.kettle.request.http.base", {
+    gradeNames: ["gpii.schema.kettle.request.http"],
+    mergePolicy: {
+        // TODO: Discuss whether to move this up to gpii-json-schema and whether the "has query data" mix-in grade is worth preserving.
+        "rules.requestContentToValidate": "nomerge",
+        "rules.filterInput": "nomerge"
+    },
+    settingsSchemaPath: "%gpii-universal/build/schemas/settings-schema.json",
+    settingsSchema: "@expand:gpii.universal.solutionsRegistry.getSettingsSchema({that}.options.settingsSchemaPath)",
+    filteredSettingsSchema: "@expand:gpii.universal.solutionsRegistry.kettle.request.http.generateFilteredSchema({that}.options.settingsSchema, {request}.req, {that}.options.rules.filterInput)",
+    rules: {
+        filterInput: "{that}.options.rules.requestContentToValidate"
+    },
+    components: {
+        validationMiddleware: {
+            options: {
+                inputSchema: "{gpii.universal.solutionsRegistry.kettle.request.http.base}.options.inputSchema"
+            }
+        }
+    }
+});
+
+fluid.defaults("gpii.universal.solutionsRegistry.kettle.request.http.settings", {
+    gradeNames: ["gpii.universal.solutionsRegistry.kettle.request.http.base"],
+    rules: {
+        filterInput: {
+            "": "body.flat.contexts.gpii-default.preferences"
+        }
+    },
+    inputSchema: {
+        type: "object",
+        properties: {
+            flat: {
+                type: "object",
+                required: true,
+                properties: {
+                    contexts: {
+                        type: "object",
+                        required: true,
+                        additionalProperties: {
+                            type: "object",
+                            required: true,
+                            properties: {
+                                settings: "{that}.options.filteredSettingsSchema"
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+});
+
+fluid.defaults("gpii.universal.solutionsRegistry.kettle.request.http.preferences", {
+    gradeNames: ["gpii.universal.solutionsRegistry.kettle.request.http.base"],
+    rules: {
+        filterInput: {
+            "": {
+                transform: {
+                    type: "fluid.transforms.firstValue",
+                    // TODO: discuss how to handle contexts other than gpii-default.
+                    values: [
+                        "body.contexts.gpii-default.preferences.application",
+                        "body.contexts.gpii-default.preferences"
+                    ]
+                }
+            }
+        }
+    },
+    inputSchema: {
```

Comment:
As discussed verbally, it is probably best to factorise this task in two and derive grades whose purpose is just to store "pure schema" (schema-holding grades) and then to direct the validation at these, rather than expect the total schema consumed by a request validator to be written out inline in its own grade

Reply:
I really hoped that work could wait for the next (LSR) spike.

Reply:
Thinking about this overnight, I think we need to discuss inheritance strategies in a bit more depth, and flesh out our best ideas as a well-documented improvement to the gpii-json-schema package that will be used in the LSR work.